cloudfoundry/managers/v3appdeployers/bluegreen_strategy_v3.go

```diff
@@ -470,3 +436,48 @@ func isAppStopped(clientV3 *ccv3.Client, appGUID string) (bool, error) {
 	// All instances stopped
 	return true, err
 }
+
+func SafeAppDeletion(client ccv3.Client, appGuid string, remainingAttempts int) error {
+	// start deletion job
+	jobURL, _, err := client.DeleteApplication(appGuid)
+	if err != nil {
+		return err
+	}
+
+	// poll completion
+	err = common.PollingWithTimeout(func() (bool, error) {
+		job, _, err := client.GetJob(jobURL)
+		if err != nil {
+			return true, err
+		}
+
+		if job.State == constantV3.JobFailed {
+			return true, fmt.Errorf(
+				"App deletion failed, reason: %+v",
+				job.Errors(),
+			)
+		}
+
+		if job.State == constantV3.JobComplete {
+			return true, nil
+		}
+
+		return false, nil
+	}, 5*time.Second, 1*time.Minute)
+
+	// error handling
+	if err != nil {
+		// https://github.com/cloudfoundry/cloud_controller_ng/issues/3589 -> Delete app when bound to service fails with async service brokers -> Retry that
+		specialError, _ := regexp.MatchString("An operation for the service binding between app .* and service instance .* is in progress.", err.Error())
```

Comment:
Out of curiosity, will the API return an error matching this regardless of the service broker implementation?

Reply:
They even have a test for this at this point: https://github.com/cloudfoundry/cloud_controller_ng/blob/ee25971edd577137ecc1e7104ad343e55043de9f/spec/unit/actions/app_delete_spec.rb#L236-L248 -> whenever possible, they go async, and then they expect this message

The app delete action uses the bindings deletion mixin: https://github.com/cloudfoundry/cloud_controller_ng/blob/main/app/actions/mixins/bindings_delete.rb
The actual delete action responds here: https://github.com/cloudfoundry/cloud_controller_ng/blob/ee25971edd577137ecc1e7104ad343e55043de9f/app/actions/v3/service_binding_delete.rb#L78
If the action is not complete yet, this here is triggered: https://github.com/cloudfoundry/cloud_controller_ng/blob/ee25971edd577137ecc1e7104ad343e55043de9f/app/actions/mixins/bindings_delete.rb#L23
Which resolves to https://github.com/cloudfoundry/cloud_controller_ng/blob/main/app/actions/app_delete.rb#L106, where the message is hardcoded